lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c

```diff
@@ -2059,6 +2059,326 @@ int s1ap_mme_handle_handover_request_ack(
+
+  // cause (mandatory)
+  hpf_ie = (S1ap_HandoverPreparationFailureIEs_t*) calloc(
+      1, sizeof(S1ap_HandoverPreparationFailureIEs_t));
+  hpf_ie->id            = S1ap_ProtocolIE_ID_id_Cause;
+  hpf_ie->criticality   = S1ap_Criticality_ignore;
+  hpf_ie->value.present = S1ap_HandoverPreparationFailureIEs__value_PR_Cause;
+  s1ap_mme_set_cause(&hpf_ie->value.choice.Cause, cause_type, cause_value);
+  ASN_SEQUENCE_ADD(&out->protocolIEs.list, hpf_ie);
+
+  // Construct the PDU and send message
+  if (s1ap_mme_encode_pdu(&out_pdu, &buffer_p, &length) < 0) {
+    err = 1;
+  }
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_HandoverPreparationFailure, out);
+  if (err) {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+  bstring b = blk2bstr(buffer_p, length);
+  free(buffer_p);
+
+  OAILOG_DEBUG(
+      LOG_S1AP,
+      "send HANDOVER PREPARATION FAILURE for mme_ue_s1ap_id " MME_UE_S1AP_ID_FMT
+      "\n",
+      (uint32_t) mme_ue_s1ap_id);
+
+  s1ap_mme_itti_send_sctp_request(
+      &b, ue_ref_p->sctp_assoc_id, ue_ref_p->sctp_stream_send, mme_ue_s1ap_id);
+
+  OAILOG_FUNC_RETURN(LOG_S1AP, RETURNok);
+}
+
+int s1ap_mme_handle_handover_cancel(
+    s1ap_state_t* state, const sctp_assoc_id_t assoc_id,
+    const sctp_stream_id_t stream, S1ap_S1AP_PDU_t* pdu) {
+  S1ap_HandoverCancel_t* container = NULL;
+  S1ap_HandoverCancelIEs_t* ie     = NULL;
+  S1ap_S1AP_PDU_t out_pdu          = {0};
+  S1ap_HandoverCancelAcknowledge_t* out;
+  S1ap_HandoverCancelAcknowledgeIEs_t* hca_ie = NULL;
+  ue_description_t* ue_ref_p                  = NULL;
+  mme_ue_s1ap_id_t mme_ue_s1ap_id             = INVALID_MME_UE_S1AP_ID;
+  enb_ue_s1ap_id_t enb_ue_s1ap_id             = INVALID_ENB_UE_S1AP_ID;
+  S1ap_Cause_PR cause_type;
+  long cause_value;
+  uint8_t* buffer_p = NULL;
+  uint8_t err       = 0;
+  uint32_t length   = 0;
+
+  OAILOG_FUNC_IN(LOG_S1AP);
+
+  container = &pdu->choice.initiatingMessage.value.choice.HandoverCancel;
+
+  // get the mandantory IEs
+  // MME_UE_S1AP_ID
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverCancelIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_MME_UE_S1AP_ID, true);
+  if (ie) {
+    mme_ue_s1ap_id = ie->value.choice.MME_UE_S1AP_ID;
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // eNB_UE_S1AP_ID
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverCancelIEs_t, ie, container,
+      S1ap_ProtocolIE_ID_id_eNB_UE_S1AP_ID, true);
+  // eNB UE S1AP ID is limited to 24 bits
+  if (ie) {
+    enb_ue_s1ap_id = (enb_ue_s1ap_id_t)(
+        ie->value.choice.ENB_UE_S1AP_ID & ENB_UE_S1AP_ID_MASK);
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+
+  // Grab the Cause Type and Cause Value
+  S1AP_FIND_PROTOCOLIE_BY_ID(
+      S1ap_HandoverCancelIEs_t, ie, container, S1ap_ProtocolIE_ID_id_Cause,
+      true);
+  if (ie) {
+    cause_type = ie->value.choice.Cause.present;
+  } else {
+    OAILOG_FUNC_RETURN(LOG_S1AP, RETURNerror);
+  }
+  switch (cause_type) {
```

Comment:
considering this is already reused in this PR twice, actually it is better to factor out the setting of the cause_value.